nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/QueryIoTDBRecord.java

```diff
@@ -0,0 +1,219 @@
+            if ( (incomingFlowFile != null ? incomingFlowFile.getSize() : 0) == 0 ) {
+                if ( context.getProperty(IOTDB_QUERY).isSet() ) {
+                    query = context.getProperty(IOTDB_QUERY).evaluateAttributeExpressions(incomingFlowFile).getValue();
+                } else {
+                    String message = "FlowFile query is empty and no scheduled query is set";
+                    getLogger().error(message);
+                    incomingFlowFile = processSession.putAttribute(incomingFlowFile, "iotdb.error.message", message);
+                    processSession.transfer(incomingFlowFile, REL_FAILURE);
+                    return;
+                }
+            } else {
+                try {
+                    query = getQuery(processSession, incomingFlowFile);
+                } catch(IOException ioe) {
+                    final String message = String.format("Reading Query from FlowFile failed %s", incomingFlowFile);
+                    throw new ProcessException(message, ioe);
+                }
+            }
+            outgoingFlowFile = incomingFlowFile;
+        } else {
+            outgoingFlowFile = processSession.create();
+            query = context.getProperty(IOTDB_QUERY).evaluateAttributeExpressions(outgoingFlowFile).getValue();
+        }
+
+        try {
+            SessionDataSet dataSet = session.get().executeQueryStatement(query);
+            List<Map<String,Object>> result = new ArrayList<>();
+            dataSet.setFetchSize(fetchSize);
+            List<String> fieldType = dataSet.getColumnTypes();
+            List<String> fieldNames = dataSet.getColumnNames();
+            List<RecordField> recordFields = new ArrayList<>();
+            for (int i = 0; i < fieldNames.size(); i++) {
+                recordFields.add(new RecordField(fieldNames.get(i), getType(fieldType.get(i)).getDataType()));
+            }
+            final RecordSchema schema = new SimpleRecordSchema(recordFields);
+            OutputStream outputStream = processSession.write(outgoingFlowFile);
+            RecordSetWriter resultSetWriter =
+                    recordSetWriterFactory.createWriter(getLogger(),schema,outputStream,outgoingFlowFile);
+            List<Record> records = new ArrayList<>();
+            while (dataSet.hasNext()) {
+                final Map<String, Object> map = new LinkedHashMap<>();
+                RowRecord rowRecord = dataSet.next();
+                map.put(fieldNames.get(0), rowRecord.getTimestamp()); //Put the timestamp
+                List<Field> fields = rowRecord.getFields();
+                for(int i = 0; i< fieldNames.size() - 1; i++ ){ //Put the rest data
+                    TSDataType dataType = fields.get(i).getDataType();
+                    map.put(fieldNames.get(i+1), fields.get(i).getObjectValue(dataType));
+                }
+                result.add(map);
+                Record record = new MapRecord(schema, map);
```

Comment:
Extracting the conversion from IoTDB `RowRecord` to NiFi `Record` would help simplify the overall method.